File: mobile/modules/vibe-chat-native/ios/VibeNativeCallModule.swift
import ExpoModulesCore

public class VibeNativeCallModule: Module {
  deinit {
    VibeNativeCallUiCoordinator.shared.detach(module: self)
  }

  public func definition() -> ModuleDefinition {
    Name("VibeNativeCall")

    OnCreate {
      NSLog("[VibeNativeCall][UiModule] OnCreate attach")
      VibeNativeCallUiCoordinator.shared.attach(module: self)
    }

    Events("onCallUiEvent")

    Function("isSupported") {
      true
    }

    Function("supportsInAppUi") {
      true
    }

    Function("setNativeEngineConfig") { (payload: [String: Any]) in
      VibeNativeCallEngine.shared.configure(payload)
    }

    Function("getNativeEngineConfig") {
      VibeNativeCallEngine.shared.getConfig()
    }

    Function("getNativeEngineStatus") {
      VibeNativeCallEngine.shared.getStatus()
    }

    Function("getNativeIceConfig") {
      VibeNativeCallEngine.shared.getIceConfig()
    }

    Function("getNativeSignalingJournal") {
      VibeNativeCallEngine.shared.getSignalingJournal()
    }

    Function("clearNativeSignalingJournal") {
      VibeNativeCallEngine.shared.clearSignalingJournal()
    }

    Function("nativeRefreshTurnConfig") {
      VibeNativeCallEngine.shared.refreshTurnConfig(force: true)
    }

    Function("nativeStartOutgoingCall") { (payload: [String: Any]) in
      VibeNativeCallEngine.shared.startOutgoing(payload)
    }

    Function("nativeAcceptIncomingCall") { (payload: [String: Any]) in
      VibeNativeCallEngine.shared.acceptIncoming(payload)
    }

    Function("nativeHandleSignal") { (payload: [String: Any]) in
      VibeNativeCallEngine.shared.handleSignal(payload)
    }

    Function("nativeEndCall") { (payload: [String: Any]) in
      VibeNativeCallEngine.shared.endCall(payload)
    }

    Function("drainPendingEvents") {
      let events = VibeNativeCallStore.shared.drainEvents()
      NSLog("[VibeNativeCall][Module] drainPendingEvents count=%d", events.count)
      return events
    }

    Function("getPushTokens") {
      let tokens = VibeNativeCallStore.shared.getPushTokens()
      let hasVoip = ((tokens["voip"] as? String)?.isEmpty == false)
      let hasApns = ((tokens["apns"] as? String)?.isEmpty == false)
      NSLog("[VibeNativeCall][Module] getPushTokens hasVoip=%@ hasApns=%@", hasVoip ? "true" : "false", hasApns ? "true" : "false")
      return tokens
    }

    Function("clearIncomingCallUi") { (payload: [String: Any]) in
      let callId = (payload["callId"] as? String) ?? (payload["call_id"] as? String)
      VibeNativeCallManager.shared.clearIncomingCallUi(callId: callId)
    }

    Function("setCallUiState") { (payload: [String: Any]) in
      VibeNativeCallUiCoordinator.shared.captureReactBridgeObject(self.appContext?.reactBridge)
      if let legacyProxy = self.appContext?.legacyModulesProxy as? NSObject {
        let legacyBridge = legacyProxy.value(forKey: "bridge") as AnyObject?
        VibeNativeCallUiCoordinator.shared.captureReactBridgeObject(legacyBridge)
      }
      let mode = (payload["mode"] as? String) ?? "hidden"
      let visible = (payload["visible"] as? Bool) ?? (mode != "hidden")
      let callId = (payload["callId"] as? String) ?? (payload["call_id"] as? String) ?? "-"
      let status = (payload["callStatus"] as? String) ?? "-"
      NSLog(
        "[VibeNativeCall][UiModule] setCallUiState mode=%@ visible=%@ callId=%@ status=%@",
        mode, visible ? "true" : "false", callId, status
      )
      VibeNativeCallUiCoordinator.shared.setState(payload)
    }

    Function("hideCallUi") {
      NSLog("[VibeNativeCall][UiModule] hideCallUi")
      VibeNativeCallUiCoordinator.shared.hide()
    }

    OnDestroy {
      NSLog("[VibeNativeCall][UiModule] OnDestroy detach")
      VibeNativeCallUiCoordinator.shared.detach(module: self)
    }
  }

  func emitCallUiEvent(_ payload: [String: Any]) {
    NSLog(
      "[VibeNativeCall][UiModule] emitCallUiEvent type=%@",
      (payload["type"] as? String) ?? "-"
    )
    sendEvent("onCallUiEvent", payload)
  }
}

File: mobile/modules/vibe-chat-native/android/src/main/java/expo/modules/vibechatnative/VibeNativeCallModule.kt
package expo.modules.vibechatnative

import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.vibechatnative.notifications.VibeIncomingCallNotification
import expo.modules.vibechatnative.notifications.VibeNativeCallStore
import expo.modules.vibechatnative.notifications.VibeNativeCallUiBridge

class VibeNativeCallModule : Module() {
  init {
    VibeNativeCallUiBridge.attachModule(this)
  }

  override fun definition() = ModuleDefinition {
    Name("VibeNativeCall")

    Events("onCallUiEvent")

    Function("isSupported") {
      true
    }

    Function("supportsInAppUi") {
      true
    }

    Function("setNativeEngineConfig") { payload: Map<String, Any?> ->
      val context = appContext.reactContext ?: appContext.currentActivity?.applicationContext ?: return@Function emptyMap<String, Any?>()
      NativeCallEngine.configure(context, payload)
    }

    Function("getNativeEngineConfig") {
      val context = appContext.reactContext ?: appContext.currentActivity?.applicationContext ?: return@Function emptyMap<String, Any?>()
      NativeCallEngine.getConfig(context)
    }

    Function("getNativeEngineStatus") {
      NativeCallEngine.getStatus()
    }

    Function("getNativeIceConfig") {
      val context = appContext.reactContext ?: appContext.currentActivity?.applicationContext ?: return@Function emptyMap<String, Any?>()
      NativeCallEngine.getIceConfig(context)
    }

    Function("getNativeSignalingJournal") {
      NativeCallEngine.getSignalingJournal()
    }

    Function("clearNativeSignalingJournal") {
      NativeCallEngine.clearSignalingJournal()
    }

    Function("nativeRefreshTurnConfig") {
      val context = appContext.reactContext ?: appContext.currentActivity?.applicationContext ?: return@Function emptyMap<String, Any?>()
      NativeCallEngine.refreshTurnConfig(context, force = true)
    }

    Function("nativeStartOutgoingCall") { payload: Map<String, Any?> ->
      NativeCallEngine.startOutgoing(payload)
    }

    Function("nativeAcceptIncomingCall") { payload: Map<String, Any?> ->
      NativeCallEngine.acceptIncoming(payload)
    }

    Function("nativeHandleSignal") { payload: Map<String, Any?> ->
      NativeCallEngine.handleSignal(payload)
    }

    Function("nativeEndCall") { payload: Map<String, Any?> ->
      NativeCallEngine.endCall(payload)
    }

    Function("drainPendingEvents") {
      VibeNativeCallStore.drainEvents(appContext.reactContext ?: appContext.currentActivity?.applicationContext ?: return@Function emptyList<Map<String, Any?>>())
    }

    Function("getPushTokens") {
      val context = appContext.reactContext ?: appContext.currentActivity?.applicationContext
      val fcm = context?.let { VibeNativeCallStore.getFcmToken(it) }
      mapOf(
        "platform" to "android",
        "fcm" to fcm,
      )
    }

    Function("clearIncomingCallUi") { payload: Map<String, Any?> ->
      val context = appContext.reactContext ?: appContext.currentActivity?.applicationContext ?: return@Function
      val callId = (payload["callId"] ?: payload["call_id"])?.toString()
      VibeIncomingCallNotification.cancelIncomingCall(context, callId)
    }

    Function("setCallUiState") { payload: Map<String, Any?> ->
      val context = appContext.reactContext ?: appContext.currentActivity?.applicationContext ?: return@Function
      val mode = (payload["mode"] as? String) ?: "hidden"
      val visible = (payload["visible"] as? Boolean)
        ?: (mode != "hidden")
      Log.d(
        "VibeNativeCall",
        "Module.setCallUiState mode=$mode visible=$visible callId=${payload["callId"] ?: payload["call_id"]} status=${payload["callStatus"]} hasActivity=${appContext.currentActivity != null}"
      )
      VibeNativeCallUiBridge.setState(payload)
      if (visible) {
        VibeNativeCallUiBridge.present(context)
      } else {
        VibeNativeCallUiBridge.hide()
      }
    }

    Function("hideCallUi") {
      Log.d("VibeNativeCall", "Module.hideCallUi")
      VibeNativeCallUiBridge.hide()
    }

    OnDestroy {
      VibeNativeCallUiBridge.detachModule(this@VibeNativeCallModule)
    }
  }

  internal fun emitCallUiEvent(payload: Map<String, Any?>) {
    Log.d("VibeNativeCall", "Module.emitCallUiEvent type=${payload["type"]} callId=${payload["callId"] ?: payload["call_id"]}")
    try {
      sendEvent("onCallUiEvent", payload)
    } catch (_: Throwable) {
      Log.w("VibeNativeCall", "Module.emitCallUiEvent failed listenerMissing=true")
    }
  }
}
